### PR TITLE
Release profiler

### PR DESCRIPTION
**Pull request by perunt**
Usually performance issues are profiled in debug builds, which could lead to false positives (e.g. JS Garbage Collectors, Hermes-specific debug markers, or other LLDB hangs).

react-native-release-profiler allows you to profile your app in release builds, both in a local environment to profile specific issues, as well as in a production environment to spot regressions or collect performance samples from a wider variety of user devices.

This menu should be available in production. That's why I removed `isDevelopment` checks from `src/components/TestToolsModal.tsx`

After you've captured the performance trace, you can download and process it from your phone to your PC
On Android:
```npx react-native-release-profiler --fromDownload --appId <your appId>```

On IOS:
```npx react-native-release-profiler --local <path to profile>```

Also, after capturing metrics, you are able to share them. In the sharing process, you would see two files
1. Trace file 
2. App info file which looks like:
```
const appData = {
    appVersion: "1.0.0",
    environment: 'production',
    platform: "IOS",
    totalMemory: "3GB",
    usedMemory: "300MB"
}
```

### Fixed Issues

$ https://expensify.slack.com/archives/C05LX9D6E07/p1707759274500039?thread_ts=1707758148.919829&cid=C05LX9D6E07
PROPOSAL:

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Build the iOS/Android app in production mode.
2. Tap any area of the app using four fingers simultaneously.
3. Verify that you see a modal screen with only the "Use Profiling" option; no other options should be present.
4. Turn on the switcher
5. Wait a few seconds and turn off the switcher
6. You will see the 'Share' button appears 
7. Click on the button and share the trace.
8. Check if the trace was shared

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/48593211/e5655987-bd7a-4445-9df7-4258d488c0b9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/48593211/93685692-0b42-4fb2-80d6-17548e262a56

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

**Comment by situchan** [comment]
I can help if requires C+ review

**Comment by srikarparsi** [comment]
Hi, yes that would be great thanks @situchan

**Review by situchan — COMMENTED** [review]
I see you marked as draft. Please request my review when ready. Thanks

**Comment by perunt** [comment]
@situchan, now it's good to go.
I've just added a symbolication and updated README for better understanding.

**Comment by srikarparsi** [comment]
Hey @situchan, have you had a chance to look at this?

**Comment by situchan** [comment]
have been testing. @perunt please resolve conflict when you get a chance

**Comment by hannojg** [comment]
@situchan I quickly fixed the conflicts

**Comment by srikarparsi** [comment]
Started an Ad Hoc build

File: src/components/ProfilingToolMenu/index.native.tsx
Comment (hannojg) [reply src/components/ProfilingToolMenu/index.native.tsx:149]: Good question. Not entirely sure, but as its a hidden dev menu I feel like it is?

File: src/components/ProfilingToolMenu/index.native.tsx
Comment (situchan) [inline src/components/ProfilingToolMenu/index.native.tsx:146]: Not a blocker but bottom section title is so closer to top section

Screenshot 2024-02-22 at 12 56 35 AM

**Comment by situchan** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details open>
<summary>Android: Native</summary>


https://github.com/Expensify/App/assets/108292595/53134415-59e7-4d17-9e4f-733f4b9254b8



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details open>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/108292595/8acb0640-cf1e-4588-b08d-72f82712d868


https://github.com/Expensify/App/assets/108292595/7ec2ac93-58eb-464f-a622-149a024e33c8




</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Comment by situchan** [comment]
How to see `Profile_trace_for_1.4.43-7.cpuprofile`?

File: package.json
Comment (situchan) [inline package.json:331]: undo this change

**Review by situchan — APPROVED** [review]

**Comment by situchan** [comment]
Here's readme for review:
https://github.com/margelo/expensify-app-fork/tree/perunt/release-profiler?tab=readme-ov-file#release-profiler

**Comment by situchan** [comment]
Let's just address https://github.com/Expensify/App/pull/36414#discussion_r1498178989

File: package.json
Comment (hannojg) [reply package.json]: Updated!

**Comment by hannojg** [comment]
Adressed blocking comment (can't really comment on the other questions, I leave that up for @perunt to answer!)

File: package.json
Comment (situchan) [inline package.json:331]: still having space

**Comment by situchan** [comment]
@perunt these are real blockers

Screenshot 2024-02-22 at 1 55 29 AM

**Comment by perunt** [comment]
@situchan @srikarparsi typescript issues came from the main, so now, after the merge, it should be fine.
Also, just to be sure that you double-checked this test step `3. Verify that you see a modal screen with only the "Use Profiling" option; no other options should be present.` as we don't want to get the rest of the features in production

**Comment by situchan** [comment]
> @situchan @srikarparsi typescript issues came from the main, so now, after the merge, it should be fine. Also, just to be sure that you double-checked this test step `3. Verify that you see a modal screen with only the "Use Profiling" option; no other options should be present.` as we don't want to get the rest of the features in production

yes, I already confirmed in adhoc build

File: package.json
Comment (perunt) [reply package.json]: done

**Comment by perunt** [comment]
> How to see Profile_trace_for_1.4.43-7.cpuprofile?
Edit: I traced using chrome://tracing/

@situchan, with `https://www.speedscope.app/,` you will see more details. Also, you would notice that anonymous function now has a trace
Screenshot 2024-02-22 at 17 04 23

**Comment by situchan** [comment]
> @situchan @srikarparsi, if it tested well, when can we expect to get this merged?

🟢 from me

**Comment by srikarparsi** [comment]
Changes look good to me as well, waiting on [confirmation](https://expensify.slack.com/archives/C03TQ48KC/p1708622314441729) about adding new modules to package.json before merging.

**Comment by srikarparsi** [comment]
Hey @situchan and @hoangzinh, we have to get approval before adding libraries to package.json so I created two new issues for the two packages, react-native-share and react-native-release-profiler.

[react-native-share](https://github.com/Expensify/App/issues/37104) - @situchan or @hoangzinh could you please help with the last three questions. You can leave a comment in the GH issue and I can update the body.

**Comment by perunt** [comment]
@srikarparsi I've added some comments to the issues above. I hope it will help push it forward

**Comment by srikarparsi** [comment]
Hey! Could someone clarify why we're using react-native-share instead of the regular share from react native?

**Comment by perunt** [comment]
@srikarparsi, any updates on this? Can I somehow push it forward?

**Comment by srikarparsi** [comment]
Hey! Still waiting on these two issues ([1](https://github.com/Expensify/App/issues/37104), [2](https://github.com/Expensify/App/issues/37105)) I just bumped up the priority so they should hopefully be done soon.

**Comment by srikarparsi** [comment]
Hey @perunt, have you had a chance to look at this comment [here](https://github.com/Expensify/App/issues/37104#issuecomment-1969798203)

**Comment by perunt** [comment]
@srikarparsi added my comments

File: src/components/ProfilingToolMenu/index.native.tsx
Comment (roryabraham) [inline src/components/ProfilingToolMenu/index.native.tsx:135]: can you explain what this line does? Does it just open the native file in the default browser?

Comment (roryabraham) [reply src/components/ProfilingToolMenu/index.native.tsx:135]: oh, I see. This is opening the native share menu so you can do whatever you want with the file:

- save it to the filesystem
- email it
- send it in slack
- whatever else

**Comment by srikarparsi** [comment]
Awesome thanks @perunt! Also, do you think you could fix the conflicts whenever you get a chance. I think we should get approval for `react-native-share` and be able to merge by tomorrow.

**Comment by perunt** [comment]
@srikarparsi  done.
Can I speed it up somehow?

**Comment by srikarparsi** [comment]
Hey @perunt, we're still waiting on approval for react-native-share. We can merge as soon as that's done, hopefully today or tomorrow.

**Review by srikarparsi — COMMENTED** [review]
lgtm, also got approval for both libraries being added

**Review by srikarparsi — APPROVED** [review]

**Pull request merged**
